src/libc/gen/FreeBSD/getmntinfo.3
.\" Copyright (c) 1989, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)getmntinfo.3	8.1 (Berkeley) 6/9/93
.\" $FreeBSD: src/lib/libc/gen/getmntinfo.3,v 1.13 2007/01/09 00:27:54 imp Exp $
.\"
.Dd April 12, 2017
.Dt GETMNTINFO 3
.Os
.Sh NAME
.Nm getmntinfo
.Nm getmntinfo_r_np
.Nm getmntinfo64
.Nd get information about mounted file systems
.Sh SYNOPSIS
.In sys/param.h
.In sys/ucred.h
.In sys/mount.h
.Ft int
.Fn getmntinfo "struct statfs **mntbufp" "int flags"
.Ft int
.Fn getmntinfo_r_np "struct statfs **mntbufp" "int flags"
.Sh TRANSITIIONAL SYNOPSIS (NOW DEPRECATED)
.Ft int
.br
.Fn getmntinfo64 "struct statfs64 **mntbufp" "int flags" ;
.Sh DESCRIPTION
The
.Fn getmntinfo
function
returns an array of
.Ft statfs
structures describing each currently mounted file system (see
.Xr statfs 2 ) .
As
.Xr statfs 2
indicates, the structure is defined differently depending on
whether the macro _DARWIN_FEATURE_64_BIT_INODE is defined (see
.Xr stat 2
for more information on this macro).
.Pp
The
.Fn getmntinfo
and
.Fn getmntinfo_r_np
functions
pass their
.Fa flags
argument transparently to
.Xr getfsstat 2 .
.Pp
The
.Fn getmntinfo
function maintains ownership of the results buffer it allocates,
and may overwrite or free this buffer in subsequent calls to
.Fn getmntinfo .
For this reason,
.Fn getmntinfo
is not thread-safe.
.Pp
The
.Fn getmntinfo_r_np
function is a thread-safe equivalent of
.Fn getmntinfo
that allocates a new results buffer on every call and transfers ownership
of this buffer to the caller.
The caller is responsible for freeing this memory with
.Xr free 3 .
.Sh RETURN VALUES
On successful completion,
.Fn getmntinfo
and
.Fn getmntinfo_r_np
return a count of the number of elements in the array.
The pointer to the array is stored into
.Fa mntbufp .
.Pp
If an error occurs, zero is returned and the external variable
.Va errno
is set to indicate the error.
The
.Fn getmntinfo
function may modify the
.Fa mbtbufp
pointer even in the case of an error.
In this situation, callers should consider any previous information
returned by
.Fn getmntinfo
to be lost.
The
.Fn getmntinfo_r_np
function will not modify the
.Fa mntbufp
pointer in the case of an error.
.Sh ERRORS
The
.Fn getmntinfo
and
.Fn getmntinfo_r_np
functions
may fail and set errno for any of the errors specified for the library
routines
.Xr getfsstat 2
or
.Xr malloc 3 .
.Sh TRANSITIONAL DESCRIPTION (NOW DEPRECATED)
The
.Fn getmntinfo64
routine is equivalent to its corresponding non-64-suffixed routine,
when 64-bit inodes are in effect.
It was added before there was support for the symbol variants, and so is
now deprecated.
Instead of using it, set the
.Dv _DARWIN_USE_64_BIT_INODE
macro before including header files to force 64-bit inode support.
.Pp
The
.Ft statfs64
structure used by this deprecated routine is the same as the
.Ft statfs
structure when 64-bit inodes are in effect.
.Sh SEE ALSO
.Xr getfsstat 2 ,
.Xr mount 2 ,
.Xr stat 2 ,
.Xr statfs 2 ,
.Xr mount 8
.Sh HISTORY
The
.Fn getmntinfo
function first appeared in
.Bx 4.4 .
The
.Fn getmntinfo_r_np
function first appeared in macOS 10.13.
.Sh BUGS
The
.Fn getmntinfo
function writes the array of structures to an internal static object
and returns
a pointer to that object.
Subsequent calls to
.Fn getmntinfo
will modify the same object.
.Pp
The memory allocated by
.Fn getmntinfo
cannot be
.Xr free 3 Ns 'd
by the application.


src/libc/gen/FreeBSD/getmntinfo.c
/*
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if defined(LIBC_SCCS) && !defined(lint)
static char sccsid[] = "@(#)getmntinfo.c	8.1 (Berkeley) 6/4/93";
#endif /* LIBC_SCCS and not lint */
#include <sys/cdefs.h>
__FBSDID("$FreeBSD: src/lib/libc/gen/getmntinfo.c,v 1.5 2007/01/09 00:27:54 imp Exp $");

#include <sys/param.h>
#include <sys/ucred.h>
#include <sys/mount.h>
#include <errno.h>
#include <stdlib.h>

struct getmntinfo_vars {
	struct statfs *mntbuf;
	int mntsize;
	long bufsize;
};

/*
 * Return information about mounted filesystems.
 */
static int
getmntinfo_internal(struct getmntinfo_vars *vars, int flags)
{

	if (vars->mntsize <= 0 &&
	    (vars->mntsize = getfsstat(0, 0, MNT_NOWAIT)) < 0) {
		return (0);
	}
	if (vars->bufsize > 0 &&
	    (vars->mntsize =
	     getfsstat(vars->mntbuf, vars->bufsize, flags)) < 0) {
		return (0);
	}
	while (vars->bufsize <= vars->mntsize * sizeof(struct statfs)) {
		if (vars->mntbuf) {
			free(vars->mntbuf);
		}
		vars->bufsize = (vars->mntsize + 1) * sizeof(struct statfs);
		if ((vars->mntbuf =
		     (struct statfs *)malloc(vars->bufsize)) == 0) {
			return (0);
		}
		if ((vars->mntsize =
		     getfsstat(vars->mntbuf, vars->bufsize, flags)) < 0) {
			return (0);
		}
	}
	return (vars->mntsize);
}

/* Legacy version that keeps the buffer around. */
int
getmntinfo(struct statfs **mntbufp, int flags)
{
	static struct getmntinfo_vars vars;
	int rv;

	rv = getmntinfo_internal(&vars, flags);
	/* Unconditional assignment matches legacy behavior. */
	*mntbufp = vars.mntbuf;
	return (rv);
}

/* Thread-safe version where the caller owns the newly-allocated buffer. */
int
getmntinfo_r_np(struct statfs **mntbufp, int flags)
{
	struct getmntinfo_vars vars = { 0 };
	int rv, save_errno;

	if ((rv = getmntinfo_internal(&vars, flags)) != 0) {
		*mntbufp = vars.mntbuf;
	} else {
		save_errno = errno;
		free(vars.mntbuf);
		errno = save_errno;
	}
	return (rv);
}
